### DISCUSSION TITLE
Can't read dxf

### DISCUSSION THREAD

**Discussion by jap2024** [mozman/ezdxf#1025] [Q&A]
I am trying to read a dxf file (which I could read smoothly years ago), but now I get this error:
AttributeError: 'LWPolyline' object has no attribute 'add'

I think the 'add' refers to the hex handle ADD, and only ocurrence in the file looks normal code 5:
```
LWPOLYLINE
  5
ADD
100
AcDbEntity
  8
ASSEMBLY_OUTLINES_TOP
100
AcDbPolyline
 90
        2
 70
         0
 43
0.0
 10
99.814699
 20
89.796084
 10
99.566459
 20
90.254952
  0
LWPOLYLINE
  5
ADE
100
AcDbEntity
  8
ASSEMBLY_OUTLINES_TOP
100
AcDbPolyline
 90
        2
 70
         0
 43
0.0
 10
99.814699
 20
89.796084
 10
99.809542
 20
89.790928
  0
```

Rest of ocurrences are 1ADD, 2ADD, etc. Again hex handles. Anyway other files with this same 'ADD' handle can be read properly.

```
$ACADVER
  1
AC1014
```

```
# packages in environment at C:\ProgramData\anaconda3:
#
# Name                    Version                   Build  Channel
ezdxf                     0.14.2                     py_0    conda-forge
```

Could you please help to overcome this issue?

Thanks in advance and best regards.

**Comment by mozman** [comment]
Hi!

First: `ezdxf` v0.14.2 is very old (released 2020-10-18) the current version is v1.1.4.

Second: No, the error message says something completely different, your script is accessing an attribute that doesn't exist.

- I need to see your script.
- The code should be executable.
- I need the DXF file to reproduce the problem.
- Upload DXF files in a zip-archive.

**Comment by jap2024** [comment]
Hi mozman,
Thanks for your response.
This is the code I used to read the dxf file (https://ezdxf.readthedocs.io/en/stable/tutorials/getting_data.html). I tried also with the recover approach (https://ezdxf.readthedocs.io/en/stable/drawing/recover.html#module-ezdxf.recover).

```
import sys
import ezdxf
from ezdxf import recover

#MAIN--------------------------------
# try:
#     dwg = ezdxf.readfile("test.dxf")
# except IOError:
#     print(f"Not a DXF file or a generic I/O error.")
#     sys.exit(1)
# except ezdxf.DXFStructureError:
#     print(f"Invalid or corrupted DXF file.")
#     sys.exit(2)

#-------------------------------------+  
try:
    doc, auditor = recover.readfile("test.dxf")
except IOError:
    print(f'Not a DXF file or a generic I/O error.')
    sys.exit(1)
except ezdxf.DXFStructureError:
    print(f'Invalid or corrupted DXF file.')
    sys.exit(2)

# DXF file can still have unrecoverable errors, but this is maybe just
# a problem when saving the recovered DXF file.
if auditor.has_errors:
    auditor.print_error_report()

    
#-------------------------------------+   
print ("DXF version : {}".format(dwg.dxfversion))

msp = dwg.modelspace()
print ("DXF number of entities : {}".format(len(msp)))
```


And this is the output I get:

  ```
File C:\ProgramData\anaconda3\Lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\python_projects\read_test_dxf.py:27
    doc, auditor = recover.readfile("test.dxf")

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\recover.py:54 in readfile
    doc, auditor = read(fp, errors=errors)

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\recover.py:82 in read
    doc._load_section_dict(recover_tool.section_dict)

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\document.py:361 in _load_section_dict
    self._finalize_setup()

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\document.py:138 in _finalize_setup
    self.mline_styles = MLineStyleCollection(self)

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\entities\mline.py:273 in __init__
    self.create_required_entries()

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\entities\mline.py:277 in create_required_entries
    entity: MLineStyle = self.new('Standard')

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\entities\objectcollection.py:71 in new
    return self._new(name, dxfattribs={'name': name})

  File C:\ProgramData\anaconda3\Lib\site-packages\ezdxf\entities\objectcollection.py:78 in _new
    self.object_dict.add(name, obj)

AttributeError: 'LWPolyline' object has no attribute 'add'
```


About the ezdxf version, v0.14.2 is what I get by default. I probably used even older versions back in 2018 when I used the file very often and with no errors. I have tried to force latest ezdxf in my Anaconda install (ezdxf=v1.1.0), but conflicts were found. After 2 days solving conflicts the package was not installed. I have a 6K+ lines partial log if you are interested.

About the dxf file, sorry I can't share. I will let you know if I can find a way to trim or scramble it while still reproducing the error.

Best regards.

**Comment by mozman** [comment]
This appears to be a DXF structure error, but I can only tell for sure when I get the DXF file. I added a check in `Dictionary.get_required_dict()` to accept only `DICTIONARY` entities, which in your case is a `LWPOLYLINE` entity. That check raises a `DXFStructure` exception as it is an unrecoverable error. What application originally created the DXF file?
